src/config.js

```diff
@@ -0,0 +1,58 @@
+/* @flow */
+import requireUncached from 'require-uncached';
+
+import {createLogger} from './util/logger';
+import {UsageError} from './errors';
+
+const log = createLogger(__filename);
+
+type ApplyConfigToArgvParams = {|
+  argv: Object,
+  configObject: Object,
+  defaultValues: Object,
+|};
+
+export function applyConfigToArgv({
+  argv,
+  configObject,
+  defaultValues = {},
+}: ApplyConfigToArgvParams): Object {
+  const newArgv = {...argv};
+  for (const option in configObject) {
+    // we assume the value was set on the CLI if the default value is
+    // not the same as that on the argv object as there is a very rare chance
+    // this happening
+    const wasValueSetOnCLI = typeof(argv[option]) !== 'undefined' &&
+      (argv[option] !== defaultValues[option]);
+    if (wasValueSetOnCLI) {
+      log.debug(`Favoring CLI: ${option}=${argv[option]} over ` +
+        `configuration: ${option}=${configObject[option]}`);
+      continue;
+    }
+    // We want to include options of types string, boolean and number
+    if (argv[option] === undefined) {
```

Comment:
Sorry, I was wrong earlier about the prototype inheritance problem when using `hasOwnProperty()`. That method was the correct one to use because it's definitely possible for a *defined* option to be given a value of undefined (this is one of the biggest head scratchers in the JavaScript language but it is what it is).

Here is a patch for a test that fails with the way you have it here along with a fix:

````diff
diff --git a/src/config.js b/src/config.js
index f33206b..ee0f0d0 100644
--- a/src/config.js
+++ b/src/config.js
@@ -30,7 +30,7 @@ export function applyConfigToArgv({
       continue;
     }
     // We want to include options of types string, boolean and number
-    if (argv[option] === undefined) {
+    if (!argv.hasOwnProperty(option)) {
       log.debug(`Favoring configuration: ${option}=${configObject[option]} ` +
         `over CLI: ${option}=${argv[option]}`);
       continue;
@@ -55,4 +55,4 @@ export function loadJSConfigFile(filePath: string): Object {
       'Did you set module.exports = {...}?');
   }
   return configObject;
-}
\ No newline at end of file
+}
diff --git a/tests/unit/test.config.js b/tests/unit/test.config.js
index fae0bd8..5753c37 100644
--- a/tests/unit/test.config.js
+++ b/tests/unit/test.config.js
@@ -210,6 +210,22 @@ describe('config', () => {
       assert.strictEqual(newArgv.sourceDir, cmdLineSrcDir);
     });
 
+    it('uses configured value even when option defaults to undefined', () => {
+      const {argv, defaultValues} = makeArgv({
+        globalOpt: {
+          'source-dir': {
+            type: 'string',
+            default: undefined,
+          },
+        },
+      });
+      const configObject = {
+        sourceDir: '/configured/directory',
+      };
+      const newArgv = applyConfigToArgv({argv, configObject, defaultValues});
+      assert.strictEqual(newArgv.sourceDir, '/configured/directory');
+    });
+
     it('uses a configured number value over a falsey default', () => {
       const {argv, defaultValues} = makeArgv({
         userCmd: ['fakecommand'],
````